generate/traverse.go

```diff
@@ -0,0 +1,357 @@
+// query, and convertInputType, which builds the argument-types.
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/vektah/gqlparser/v2/ast"
+)
+
+// baseTypeForOperation returns the definition of the GraphQL type to which the
+// root of the operation corresponds, e.g. the "Query" or "Mutation" type.
+func (g *generator) baseTypeForOperation(operation ast.Operation) (*ast.Definition, error) {
+	switch operation {
+	case ast.Query:
+		return g.schema.Query, nil
+	case ast.Mutation:
+		return g.schema.Mutation, nil
+	case ast.Subscription:
+		if !allowBrokenFeatures {
+			return nil, errorf(nil, "genqlient does not yet support subscriptions")
+		}
+		return g.schema.Subscription, nil
+	default:
+		return nil, errorf(nil, "unexpected operation: %v", operation)
+	}
+}
+
+// convertOperation builds the response-type into which the given operation's
+// result will be unmarshaled.
+func (g *generator) convertOperation(
+	operation *ast.OperationDefinition,
+	queryOptions *GenqlientDirective,
+) (goType, error) {
+	name := operation.Name + "Response"
+
+	if def, ok := g.typeMap[name]; ok {
+		return nil, errorf(operation.Position, "%s defined twice:\n%s", name, def)
+	}
+
+	baseType, err := g.baseTypeForOperation(operation.Operation)
+	if err != nil {
+		return nil, errorf(operation.Position, "%v", err)
+	}
+
+	goTyp, err := g.convertDefinition(
+		name, operation.Name, baseType, operation.Position,
+		operation.SelectionSet, queryOptions)
+
+	if structType, ok := goTyp.(*goStructType); ok {
+		// Override the ordinary description; the GraphQL documentation for
+		// Query/Mutation is unlikely to be of value.
+		// TODO(benkraft): This is a bit awkward/fragile.
+		structType.Description =
+			fmt.Sprintf("%v is returned by %v on success.", name, operation.Name)
+		structType.Incomplete = false
+	}
+
+	return goTyp, err
+}
+
+var builtinTypes = map[string]string{
+	// GraphQL guarantees int32 is enough, but using int seems more idiomatic
+	"Int":     "int",
+	"Float":   "float64",
+	"String":  "string",
+	"Boolean": "bool",
+	"ID":      "string",
+}
+
+// typeName computes the name, in Go, that we should use for the given
+// GraphQL type definition.  This is dependent on its location within the query
+// (see DESIGN.md for more on why we generate type-names this way), which is
+// determined by the prefix argument; the nextPrefix result should be passed to
+// calls to typeName on any child types.
+func (g *generator) typeName(prefix string, typ *ast.Definition) (name, nextPrefix string) {
+	typeGoName := upperFirst(typ.Name)
+	if typ.Kind == ast.Enum || typ.Kind == ast.InputObject {
+		// STOPSHIP: hopefully inline this
+		// If we're an enum or an input-object, there is only one type we
+		// will ever possibly generate for this type, so we don't need any
+		// of the qualifiers.  This is especially helpful because the
+		// caller is very likely to need to reference these types in their
+		// code.
```

Comment:
I didn't see an obvious alternative place to put this. Is this comment suggesting the code be inlined in `convertField` (and ignore the place typeName is used in `convertDefinition` since enums and input objects aren't valid types to use with abstract fields)?

Reply:
Whoops, I meant to remove this before landing, but apparently I only removed it in the next diff!  Let me see if I can backport the changes from there.